### DISCUSSION TITLE
LLVM export templates

### DISCUSSION THREAD

**Discussion by sammonius** [godotengine/godot-proposals#9359] [Engine Core]
(Some context for anyone who hasn't heard of LLVM: it's bascially an optimizer made to be used in compilers. The compiler turns the code into LLVM assembly, which LLVM then optimizes and compiles into a binary.)

Since the default export templates are already compiled, it's impossible to optimize the engine for a specific game without compiling a custom export template. I think this could be fixed if the default export templates were in LLVM assembly instead, and then the engine was assembled in the editor along with the game. This would mean that unused features could be removed automatically (not just nodes, but also the funcionality within the nodes) and features could be optimized based on how they are used. It would also potentially allow GDScript to be optimized by compiling it to asembly.

One issue might be that it increases build times, but even with full optimization it would probably just take a few minutes.

**Comment by AThousandShips** [comment]
These would most likely be enormous, can you confirm these outputs aren't massive?

**Comment by sammonius** [comment]
I'll test it to make sure, but i think it won't be too big if link-time optimization is used.

**Reply by AThousandShips** [reply]
Assembly isn't linked though right? And link time optimizations are target specific isn't it? It also can't be done until the end so any changes after would be impossible

**Reply by AThousandShips** [reply]
Also this would involve integrating LLVM into the editor which isn't going to happen, LLVM is huge, see:
* https://github.com/godotengine/godot-proposals/issues/6031#issuecomment-1368409996

You'd also add a lot of overhead for the various libraries in the editor, also this won't work for exports to web or iOS which use specific compilers 

Also this wouldn't simply allow you to remove unused features either, that isn't something you can just do in the linking stage, the game isn't compiled, so you couldn't detect what's used or not

**Reply by sammonius** [reply]
Oh, ok thanks. I didn't know any of that.
